providers/k8s/resources/discovery.go

```diff
@@ -0,0 +1,735 @@
+	// it should be excluded
+	for _, ns := range f.exclude {
+		g, err := glob.Compile(ns)
+		if err != nil {
+			log.Error().Err(err).Msg("failed to compile glob")
+			return false
+		}
+		if g.Match(namespace) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func Discover(runtime *plugin.Runtime) (*inventory.Inventory, error) {
+	conn := runtime.Connection.(shared.Connection)
+
+	in := &inventory.Inventory{Spec: &inventory.InventorySpec{
+		Assets: []*inventory.Asset{},
+	}}
+
+	invConfig := conn.InventoryConfig()
+
+	res, err := runtime.CreateResource(runtime, "k8s", nil)
+	if err != nil {
+		return nil, err
+	}
+	k8s := res.(*mqlK8s)
+
+	nsFilter := NamespaceFilterOpts{}
+	if include, ok := invConfig.Options[shared.OPTION_NAMESPACE]; ok && len(include) > 0 {
+		nsFilter.include = strings.Split(include, ",")
+	}
+
+	if exclude, ok := invConfig.Options[shared.OPTION_NAMESPACE_EXCLUDE]; ok && len(exclude) > 0 {
+		nsFilter.exclude = strings.Split(exclude, ",")
+	}
+
+	// If we can discover the cluster asset, then we use that as root and build all
+	// platform IDs for the assets based on it. If we cannot discover the cluster, we
+	// discover the individual namespaces according to the ns filter and then build
+	// the platform IDs for the assets based on the namespace.
+	if slices.Contains(invConfig.Discover.Targets, DiscoveryClusters) {
+		assetId, err := conn.AssetId()
+		if err != nil {
+			return nil, err
+		}
+		in.Spec.Assets = append(in.Spec.Assets, &inventory.Asset{
+			PlatformIds: []string{assetId},
+			Name:        conn.Name(),
+			Platform:    conn.Platform(),
+			Connections: []*inventory.Config{invConfig}, // pass-in the parent connection config TODO: clone the config
+		})
+
+		assets, err := discoverAssets(conn, invConfig, assetId, k8s, nsFilter, false)
+		if err != nil {
+			return nil, err
+		}
+		in.Spec.Assets = append(in.Spec.Assets, assets...)
+	} else {
+		nss, err := discoverNamespaces(conn, invConfig, "", nsFilter)
+		if err != nil {
+			return nil, err
+		}
+
+		in.Spec.Assets = append(in.Spec.Assets, nss...)
+
+		// Discover the assets for each namespace and use the namespace platform ID as root
+		for _, ns := range nss {
+			nsFilter = NamespaceFilterOpts{include: []string{ns.Name}}
+
+			// We don't want to discover the namespaces again since we have already done this above
+			assets, err := discoverAssets(conn, invConfig, ns.PlatformIds[0], k8s, nsFilter, true)
+			if err != nil {
+				return nil, err
+			}
+			in.Spec.Assets = append(in.Spec.Assets, assets...)
+		}
+	}
+
+	return in, nil
+}
+
+func discoverAssets(
+	conn shared.Connection,
+	invConfig *inventory.Config,
+	clusterId string,
+	k8s *mqlK8s,
+	nsFilter NamespaceFilterOpts,
+	skipNsDiscovery bool,
+) ([]*inventory.Asset, error) {
+	var assets []*inventory.Asset
+	var err error
+	for _, target := range invConfig.Discover.Targets {
+		var list []*inventory.Asset
+		switch target {
+		case DiscoveryAuto:
```

Comment:
nit: in the azure discovery, i use `ifs` to see if the requested target is in the list instead of a switch statement. that way i do not call the discovery functions in multiple places (e.g. pods under DiscoveryPods and DiscoveryAuto)

Reply:
I'll change that